### PR TITLE
cigar utils tests

### PR DESCRIPTION
**Pull request by akiezun**
tests for CigarUtils + refactored methods in CigarUtils. 
Did not add tests to isValid because pull req #380 is addressing this method.

There's a potential issue in countRefBasesBasedOnCigar - it's not clear why the implementation does what it does. @amilev can you comment on the intended semantics of this method and whether it can/should use `CigarOperator.consumesReferenceBases`?

addresses #153 and #450 
@vruano please review.

### CODE REVIEW

**Comment by vruano** [comment]
Back to @akiezun. Won't be able to look at again before wednesday. It looks good once reviews are addressed and @amilev clarify semantics.

**Comment by akiezun** [comment]
assigning to @amilev to clarify the semantics of `CigarUtils.countRefBasesBasedOnCigar`.
Can / should it use `CigarOperator.consumesReferenceBases`

**Comment by akiezun** [comment]
@amilev feel free to assign back to me after clarification

File: src/main/java/org/broadinstitute/hellbender/utils/read/CigarUtils.java [hunk 1]
```diff
@@ -107,29 +109,47 @@ public static boolean isCigarValid(Cigar cigar) {
...
             switch (cigarElement.getOperator()) {
                 case M:
-                case S:
                 case D:
                 case N:
-                case H:
+                case EQ:
+                case X:
+                case S:        //TODO is this correct? CigarOperator.consumesReferenceBases is false
+                case P:        //TODO is this correct? CigarOperator.consumesReferenceBases is false
```
Comment (amilev) [inline src/main/java/org/broadinstitute/hellbender/utils/read/CigarUtils.java]: I'm not sure how P is considered, is it counted in the reference as a base, for example if I have read starting at position 10 and it cigar is 1M3P1M does the last base facing position 14 or 11? if it is the first case, then it is fine, if it is the second, then it should be treated like I

File: src/main/java/org/broadinstitute/hellbender/utils/read/CigarUtils.java [hunk 2]
```diff
@@ -107,29 +109,47 @@ public static boolean isCigarValid(Cigar cigar) {
...
+                case H:        //TODO is this correct? CigarOperator.consumesReferenceBases is false
                     result += cigarElement.getLength();
                     break;
-                case I:
+                default://I
                     break;
-                default:
-                    throw new GATKException("Unsupported cigar operator: " + cigarElement.getOperator());
```
Comment (amilev) [inline src/main/java/org/broadinstitute/hellbender/utils/read/CigarUtils.java]: I prefer to keep that check, what happens if we got some data not build by the SAM specifications? or future data with some options in the cigar that we didn't know about? What it is not good practice to catch those cases?

File: src/main/java/org/broadinstitute/hellbender/utils/read/CigarUtils.java [hunk 3]
```diff
@@ -107,29 +109,47 @@ public static boolean isCigarValid(Cigar cigar) {
...
         }
         return result;
     }
 
-    public static Cigar unclipCigar(Cigar cigar) {
-        ArrayList<CigarElement> elements = new ArrayList<>(cigar.numCigarElements());
+    /**
+     * Removes all clipping operators from the cigar.
+     */
+    public static Cigar unclipCigar(final Cigar cigar) {
```
Comment (amilev) [inline src/main/java/org/broadinstitute/hellbender/utils/read/CigarUtils.java]: it is not the right name for the function, I thought unclipping is to take the clipped bases and mark them as M instead. I would call it trimReadToUnclippedBases.

File: src/main/java/org/broadinstitute/hellbender/utils/read/CigarUtils.java [hunk 4]
```diff
@@ -137,36 +157,67 @@ public static Cigar unclipCigar(Cigar cigar) {
...
-        for (final CigarElement ce : cigar.getCigarElements()) {
-            if (ce.getOperator() == CigarOperator.N) {
-                return true;
-            }
+        if (cigar == null){
+            throw new IllegalArgumentException("null cigar");
         }
-        return false;
+        return cigar.getCigarElements().stream().anyMatch(el -> el.getOperator() == CigarOperator.N);
     }
```
Comment (amilev) [inline src/main/java/org/broadinstitute/hellbender/utils/read/CigarUtils.java]: this is nice impressing syntax

File: src/test/java/org/broadinstitute/hellbender/utils/read/CigarUtilsUnitTest.java [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+
+    @DataProvider(name = "testData_countRefBasesBasedOnCigar")
+    public Iterator<Object[]> testData_countRefBasesBasedOnCigar(final Method testMethod) {
+        final Object[][] TEST_CIGARS = {
+                {"10M", 0, 1, 10},
+                {"10M1D", 0, 1, 10},
+                {"10M1D1S", 0, 1, 10},
+                {"10M1D1N1S", 0, 1, 10},
+                {"10M1D1N1S1H", 0, 1, 10},
+                {"10M", 0, 1, 10},
```
Comment (amilev) [inline src/test/java/org/broadinstitute/hellbender/utils/read/CigarUtilsUnitTest.java]: you already have this case

File: src/test/java/org/broadinstitute/hellbender/utils/read/CigarUtilsUnitTest.java [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+                {"10M1=1X1S1H", 0, 2, 11},
+                {"10M1D2N4S", 1, 3, 3},
+                {"10M1D2N4S8H", 1, 4, 7},
+                {"10M1I2N4S", 1, 3, 2},
+                {"10M1I2N4S8H", 1, 4, 6},
+
+                //TODO are these correct? CigarOperator.consumesReferenceBases is false
+                //https://github.com/broadinstitute/hellbender/issues/450
+                {"1M1S1H1H", 1, 2, 1},
+                {"1M1P1H1H", 1, 2, 1},
```
Comment (amilev) [inline src/test/java/org/broadinstitute/hellbender/utils/read/CigarUtilsUnitTest.java]: again, I'm not sure what P should mean exactly

File: src/test/java/org/broadinstitute/hellbender/utils/read/CigarUtilsUnitTest.java [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+                {"10M1D1N1S", 0, 2, 11},
+                {"10M1D1N1S1H", 0, 2, 11},
+                {"10M1=1X1S", 0, 2, 11},
+                {"10M1=1X1S1H", 0, 2, 11},
+                {"10M1D2N4S", 1, 3, 3},
+                {"10M1D2N4S8H", 1, 4, 7},
+                {"10M1I2N4S", 1, 3, 2},
+                {"10M1I2N4S8H", 1, 4, 6},
+
+                //TODO are these correct? CigarOperator.consumesReferenceBases is false
```
Comment (amilev) [inline src/test/java/org/broadinstitute/hellbender/utils/read/CigarUtilsUnitTest.java]: I believe they are correct

**Comment by amilev** [comment]
done with my review. looks good.

File: src/main/java/org/broadinstitute/hellbender/utils/read/CigarUtils.java (hunk 3 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/read/CigarUtils.java]: done

File: src/test/java/org/broadinstitute/hellbender/utils/read/CigarUtilsUnitTest.java (hunk 5 above)
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/utils/read/CigarUtilsUnitTest.java]: thanks

File: src/main/java/org/broadinstitute/hellbender/utils/read/CigarUtils.java (hunk 2 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/read/CigarUtils.java]: i will restore it but i think we'll filter out all such reads way upstream

**Pull request merged**